### ISSUE TITLE
New macro: accepted

### ISSUE THREAD

**Issue by jodosha** [dry-rb/dry-validation#157]
``` ruby
require 'bundler/setup'
require 'dry-validation'

schema = Dry::Validation.Schema do
  required(:terms).accepted
end

result = schema.(terms: '')
puts result.success? # => false
puts result.messages # => {:terms=>["must be accepted"]}
```

The macro should expand to `bool? & true?`.

---

I've noticed that a similar macro `confirmed` isn't implemented directly on `Dry::Validation::Schema::Rule`, but on `Dry::Validation::Schema::Value`, that translates to the following syntax:

``` ruby
schema = Dry::Validation.Schema do
  required(:password).filled(min_size?: 12).confirmation
end
```

Are macros that chains `required` allowed?

**Comment by fran-worley** [comment]
@jodosha confirmation is a little different as it actually add an extra rule onto another key (implemented here: https://github.com/dry-rb/dry-validation/blob/master/lib/dry/validation/schema/value.rb#L77-L83

**Comment by jodosha** [comment]
@fran-worley Thanks, but why there isn't `required(:password).confirmed`?

**Comment by fran-worley** [comment]
I realised after I wrote that that I wasn't answering your question. I actually don't know why not. I had kinda assumed that that would work also...

**Comment by solnic** [comment]
Chaining is not supported yet, but will be in 0.8.0. It wasn't trivial to make it work so I had to postpone it. It's tricky because these objects are mutable and they gather rules, but there are cases where you don't want to add a new rule and instead you just want to return it for further composition.

In example:

``` ruby
# this needs to create a rule from `:int?` and *add it* to rules for the :age key
required(:age).value(:int?)

# this needs to create a rule inside a block (composition via `|`) and simply
# return it, and then inside `required` method we *add it* to the rule set
required(:age) { int? | some_other_pred? }
```

My goal is to enable chaining so things like that will work:

``` ruby
required(:nums).value(min_size?: 3).each(:int?)
```

**Comment by jodosha** [comment]
@solnic This ticket isn't about chaining, it's #154 instead.

---

This is about a feature request: `accepted` macro.

```ruby
schema = Dry::Validation.Schema do
  required(:terms).accepted
end
```

Do you think it makes sense to have implement this macro? Thank you! 😉

**Comment by solnic** [comment]
@jodosha I think the value of such macro would be to have a nicer default error message by default, since it's so common I believe it makes sense

**Comment by solnic** [comment]
I've re-opened this. Let's think it through if it's worth the effort /cc @timriley @AMHOL @fran-worley

**Comment by fran-worley** [comment]
@solnic all that this macro would do is essentially alias `required(:terms).value(:true?)` rather than `required(:terms).value(:accepted?)`. 

The only benefits I can see is to keep syntax inline with ActiveModel which has a similar validation and to  enable a more sensible default message than `must be true`.

Are default messages used that much? I'd hate to make an alias for the sole purpose of providing a more applicable default message for users to just override with their own.

**Comment by solnic** [comment]
I think this macro can collapse `value(:true?)` into `acceptation` just like we did with `confirmation` and then provide a default error message for `:acceptation` rule.

**Comment by fran-worley** [comment]
But as @jodosha points out, confirmation has can't be applied directly on the key (e.g. `required(:password).confirmation` so what you'd end up with in your example is having to do `required(:terms).maybe.accepted` which I don't find any better than `required(:terms).value(:true?)` other than the benefit in error message. (unless I'm missing something...)

**Comment by solnic** [comment]
It's a matter of defining this macro on `Rule` object

**Comment by fran-worley** [comment]
@solnic do we make it a macro available to both Form and Schema? As it is really aimed at forms I'd rather see it be implemented within Form in some way.

**Comment by solnic** [comment]
there's no mechanism for schema-type-specific-macros yet...we could think about it though

**Comment by fran-worley** [comment]
I stand in a similar place for 'confirmation' as to me that is quite a form specific macro.

**Comment by solnic** [comment]
We have now an API for defining custom macros, we could use it to extend `Form` with `acceptance` macro :)

**Comment by jodosha** [comment]
@solnic Do you think it's worthy do be implemented by `dry-v` or you're suggesting elsewhere (eg. Hanami)?

**Comment by solnic** [comment]
@jodosha I wouldn't be against having that in dry-v, it's super common :)

**Comment by flash-gordon** [comment]
@solnic ok, time to choose, 1.0, 2.0, ∞? :)

**Comment by solnic** [comment]
For the record, starting from 1.0.0 this type of macros should be implemented on top of the new rule API, rather than dry-schema's DSL, ie:

class NewUserContract < Dry::Validation::Contract
  params do
    required(:terms).filled(:string)
  end

  rule acceptance: :terms # or something like this
end

**Comment by solnic** [comment]
After all, this landed in master already and will be part of `1.0.0`. I added support for macros as an experimental feature for now. `:acceptance` is the first macro, here's how it looks like:

```ruby
class MyContract < Dry::Validation::Contract
  schema do
    required(:terms).value(:bool)
  end

  rule(:terms).validate(:acceptance)
end
```

Nice thing is that you can "chain" multiple macros ie `rule(:terms).validate(:check_this, :check_that, :etc)` and those will be executed as part of one rule, which means better performance.
